### DISCUSSION TITLE
Component interface best practice

### DISCUSSION THREAD

**Discussion by Burtan** [arkivanov/Decompose#149] [Q&A] [answered]
Hey,
just one short question. Why do you use interfaces for your components, when you just have one implementation?
Thanks :-)

**Comment by arkivanov** [comment, marked as answer]
Thanks for asking this question! Transferred to Discussions.

It's just a matter of preference, you can avoid interfaces if you don't like them. But there could be more than one implementation for each component.

The TodoApp example features up to three different implementations for a component:

1. Production implementation - [TodoMainComponent](https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/common/main/src/commonMain/kotlin/example/todo/common/main/integration/TodoMainComponent.kt)
2. Fake implementation for integration tests - [TodoMainFake](https://github.com/JetBrains/compose-jb/blob/master/examples/todoapp/common/root/src/commonTest/kotlin/example/todo/common/root/integration/TodoMainFake.kt)
3. Fake implementation for Compose previews - [TodoMainPreview](https://github.com/JetBrains/compose-jb/blob/97266a0ac8c0d7a8ad8d19ead1c925751a00ff1c/examples/todoapp/common/compose-ui/src/desktopMain/kotlin/example/todo/common/ui/TodoMainPreview.kt#L17)

**Reply by arkivanov** [reply]
But in general, I prefer to extract an interface for every dependency, regardless of the amount of implementations. That's just a rule of thumb for me.

**Reply by Burtan** [reply]
I didn't think of preview and tests, so that is a good call. Thanks for the fast answer!
